base.py

```diff
@@ -530,20 +544,135 @@ def group_per_episode(self, X, y = None, episode_ids = None):
+        scoring_func: str = "r2",
+        early_stopping: bool = False,
+        results_csv_path: str = "outputs/results.csv",
+        splitting_criteria: str = "CV",
+        test_indices: Union[None, List[int]] = None,
+        num_splits: int = 5,
+    ) -> pd.DataFrame:
+
+        if self.scale_data:
+            X, y = self.scalar(X, y)
+
+        if splitting_criteria.lower() == "cv":
+            cv = None
+        elif splitting_criteria.lower() == "timeseries":
+            cv = TimeSeriesSplit(n_splits=num_splits)
+        elif splitting_criteria.lower() == "grouped":
+            cv = GroupShuffleSplit(n_splits=num_splits)
+        elif splitting_criteria.lower() == "fixed":
+            if type(test_indices) != list:
+                raise ValueError("fixed split used but no test-indices provided...")
+            cv = PredefinedSplit(test_fold=test_indices)
+        else:
+            raise ValueError(
+                "Unknowing splitting criteria provided: {splitting_criteria}, should be one of [cv, timeseries, grouped]"
+            )
+
+        # early stopping only supported for learners that have a
+        # `partial_fit` method
+
+        # start mlflow auto-logging
+        mlflow.sklearn.autolog()
+
+        if search_algorithm.lower() == "bohb":
+            early_stopping = True
+
+        if any(
+            [search_algorithm.lower() in ["bohb", "bayesian", "hyperopt", "optuna"]]
+        ):
+            search = TuneSearchCV(
+                self.model,
+                params,
+                search_optimization=search_algorithm,
+                cv=cv,
+                n_trials=num_trials,
+                early_stopping=early_stopping,
+                scoring=scoring_func,
+                loggers=["csv", "tensorboard"],
+            )
+        elif search_algorithm == "grid":
+            search = GridSearchCV(
+                self.model,
+                param_grid=params,
+                refit=True,
+                cv=cv,
+                scoring=scoring_func,
+            )
+        elif search_algorithm == "random":
+            search = RandomizedSearchCV(
+                self.model,
+                param_distributions=params,
+                refit=True,
+                cv=cv,
+                scoring=scoring_func,
+            )
+        else:
+            raise NotImplementedError(
+                "Search algorithm should be one of grid, hyperopt, bohb, optuna, bayesian, or random"
+            )
+
+        with mlflow.start_run() as run:
+            search.fit(X, y)
+        self.model = search.best_estimator_
+        results_df = pd.DataFrame(search.cv_results_)
+        if not pathlib.Path(results_csv_path).parent.exists():
+            pathlib.Path(results_csv_path).parent.mkdir(exist_ok=True, parents=True)
+        logger.info(f"Saving sweeping results to {results_csv_path}")
+        results_df.to_csv(results_csv_path)
+        logger.info(f"Best hyperparams: {search.best_params_}")
+
```

Comment:
Request to output final score as well: 
`logger.info(f"Best score: {search.best_score_}")`